Clone client

mpacer opened the pull request:
This is the most basic version of the cloning endpoint for bookstore.

This adds:
- cloning functionality to the bookstore client (the `BookstoreClient` enables accessing bookstore functionality from inside a notebook)
- POST endpoint (actually does the cloning… uses Jupyter’s POST contents api to handle saving)
- GET endpoint (allows getting around xsrf issues when cloning needs to be triggered from outside of a notebook context; e.g., from commuter)
- further extensions to the minio based mock-s3 integration testing suite
- more documentation around how to setup and use that testing suite

This incorporates a lot of the work from #66 and renders it obsolete.

mpacer commented on bookstore/clone.html:
```diff
@@ -0,0 +1,56 @@
+          xhr.open("POST", url, true);
+          xhr.withCredentials = true;
+          xhr.setRequestHeader("X-XSRFToken", xsrf_token);
+          xhr.send(post_json);
```
@stormpython I'm sure I'm approaching this in a unnecessarily complicated way and would really appreciate any comments on how to simplify this (note… it doesn't work right now at all… :D)

mpacer replied:
Ah — I got it working in the latest iteration.

mpacer commented:
So the biggest open question now for me is how we should approach testing cloning. I don't want to start up selenium to test this… but there's not another good way to ensure that the cookie is being stored appropriately in the browser as far as I can think of it at least. Since we're already running in node… maybe we could leverage that? I'm not sure.

stormpython commented on ci/integration.js:
```diff
@@ -5,7 +5,7 @@ const fs = require("fs");
 const _ = require("lodash");
```
I would break out the `originalNotebook` into its own file. It makes it easier to make changes to the notebook and typically these mocked objects are usually in their own file if not folder.

mpacer replied:
Should this be in it's own PR focusing on the ci integration tests? I can do it in this one, but it's already a PR with a pretty wide scope.

mpacer replied:
NB: the other changes you asked for are stuff I'm introducing so I have no issue including those fixes in this PR.

stormpython replied:
Yes, it should go into its own PR. You can just marked this comment as resolved.

stormpython commented on ci/jupyter.js:
```diff
@@ -6,12 +6,41 @@ const { sleep } = require("./sleep");
+  initial = pieces[0].startsWith("/");
+  final = pieces[pieces.length - 1].endsWith("/");
+  const hi = [];
+  for (s of pieces) {
```
Did you know you can use `Array.map()` to achieve the same thing here? Although, I do believe a `for` loop is more performant. Here is a [blog post](https://medium.com/@ExplosionPills/map-vs-for-loop-2b4ce659fb03) that talks about using `for` vs `map`. Again, not required. But generally `map` is used in replace of a `for` loop.

mpacer replied:
Sounds good!

stormpython commented on ci/jupyter.js:
```diff
@@ -6,12 +6,41 @@ const { sleep } = require("./sleep");
 global.XMLHttpRequest = require("xmlhttprequest").XMLHttpRequest;
 const { ajax } = require("rxjs/ajax");
 
+let url_path_join = function(...pieces) {
```
In general, `const` should be used instead of `let` if the variables value never changes.

mpacer replied:
Fair enough!

mpacer commented on bookstore/clone.html:
```diff
@@ -0,0 +1,63 @@
+            s3_key: "{{ s3_key }}"
+          };
+          var clone_url = "{{ base_uri }}/api/bookstore/cloned/";
+          var r = document.cookie.match("\\b" + "_xsrf" + "=([^;]*)\\b");
```
This is the literal code that we use from jupyter notebook:

```javascript
    var _get_cookie = function (name) {
        // from tornado docs: http://www.tornadoweb.org/en/stable/guide/security.html
        var r = document.cookie.match("\\b" + name + "=([^;]*)\\b");
        return r ? r[1] : undefined;
    }
```

https://github.com/jupyter/notebook/blob/b8b66332e2023e83d2ee04f83d8814f567e01a4e/notebook/static/base/js/utils.js#L800-L804

So this logic would need to live within `getCookieByName` and would need to live in this script tag since we're not packaging this at all. Would that make you feel more confident?

willingc replied:
See PR #88

mpacer commented on ci/jupyter.js:
```diff
@@ -6,12 +6,41 @@ const { sleep } = require("./sleep");
     this.ip = config.ip || "127.0.0.1";
     this.scheme = config.scheme || "http";
     this.token = null;
+    // this.baseUrl = config.baseUrl || "";
```
I don't remember why I added it twice, I'll get rid of this one.

mpacer commented on ci/jupyter.js:
```diff
@@ -77,8 +107,9 @@ class JupyterServer {
+    const apiPath = "/api/contents/";
     const xhr = await ajax({
-      url: `${this.endpoint}/api/contents/${path}`,
+      url: url_path_join(this.endpoint, apiPath, path),
```
While developing this is constantly an issue.

Yes, its developer error with consistency around whether routes are defined with `/` at the beginning & end or not. Trying to keep track of it leaves it easy to mess up. 

I'm bringing this pattern over from the jupyter notebook server, where it is used extensively. We're already importing it for the serverside code.

This a huge quality of life improvement, trust me.

mpacer replied:
For reference to the server side code: 

```python
def url_path_join(*pieces):
    """Join components of url into a relative url
    Use to prevent double slash when joining subpath. This will leave the
    initial and final / in place
    """
    initial = pieces[0].startswith('/')
    final = pieces[-1].endswith('/')
    stripped = [s.strip('/') for s in pieces]
    result = '/'.join(s for s in stripped if s)
    if initial: result = '/' + result
    if final: result = result + '/'
    if result == '//': result = '/'
    return result
```

https://github.com/jupyter/notebook/blob/master/notebook/utils.py#L48-L61

willingc replied:
See PR #88

mpacer commented:
I think I've addressed your comments on the js side @captainsafia @stormpython. 

@willingc @MSeal would either of you be able to take a look at the python side of things?

stormpython commented on ci/utils.js:
```diff
@@ -0,0 +1,30 @@
+  //
+  // They will be readded in between words, and at the beginning and end if they were
+  // """
+  initial = pieces[0].startsWith("/");
```
missing `const`

willingc replied:
See PR #88

stormpython commented on ci/utils.js:
```diff
@@ -0,0 +1,30 @@
+  // They will be readded in between words, and at the beginning and end if they were
+  // """
+  initial = pieces[0].startsWith("/");
+  final = pieces[pieces.length - 1].endsWith("/");
```
missing `const`

willingc replied:
See PR #88

stormpython commented on ci/utils.js:
```diff
@@ -0,0 +1,30 @@
+  const hi = pieces
+    .filter(el => el !== "")
+    .map(el => el.replace(/(^[ /]+)|([/]+$)/g, "")); //[];
+  result = hi.join("/");
```
This can be rewritten to:

```javascript
const result = pieces
    .filter(el => el !== "")
    .map(el => el.replace(/(^[ /]+)|([/]+$)/g, ""))
    .join("/");
```

willingc replied:
See PR #88

MSeal commented on bookstore/client/notebook.py:
```diff
@@ -58,28 +62,48 @@ class KernelInfo:
+        self.model = {
+            "path": path,
+            "name": name,
+            "type": type,
```
we should avoid naming variables with reserved names like `type`

MSeal commented on bookstore/clone.py:
```diff
@@ -0,0 +1,112 @@
+
+        async with self.session.create_client(
+            's3',
+            aws_secret_access_key=self.bookstore_settings.s3_secret_access_key,
```
Do we force the config to be loaded into `bookstore_settings`? There's many instances where these credentials are fetch JIT from the system config loader. Likely a conversation out of scope of this PR though.

MSeal commented on bookstore/clone.py:
```diff
@@ -0,0 +1,112 @@
+        The response payload should match the standard Jupyter contents API POST response.
+        """
+        model = self.get_json_body()
+        s3_bucket = model.get("s3_bucket", "")
```
This seems wrong, an empty string bucket if the key isn't found would blow up later in the clone call. Shouldn't it fail fast here?

MSeal commented on bookstore/__init__.py:
```diff
@@ -1,3 +1,9 @@
+
+PACKAGE_DIR: str = os.path.realpath(os.path.dirname(__file__))
+
+del os
```
Maybe not expose internal paths to the init of the package as well ?

willingc replied:
Yes let's remove this.

MSeal commented on bookstore/publish.py:
```diff
@@ -0,0 +1,80 @@
+        self.session = aiobotocore.get_session()
+
+    @web.authenticated
+    async def put(self, path=''):
```
Why have the default if it will 400 anyway? Is this needed for the @web decorator?

willingc replied:
```suggestion
    async def put(self, path):
```

MSeal commented on bookstore/publish.py:
```diff
@@ -0,0 +1,80 @@
+            )
+            self.log.info("Done with published write of %s", path)
+
+        self.set_status(201)
```
should this be the last line before `finish`?

willingc commented on bookstore/client/bookstore.py:
```diff
@@ -4,18 +4,35 @@
+    def publish(self, path=None):
+        if path is None:
+            path = self.session.path
+        nb_json = self.get_contents(self.session.path)['content']
```
```suggestion
        nb_json = self.get_contents(path)['content']
```

willingc commented on bookstore/client/bookstore.py:
```diff
@@ -4,18 +4,35 @@
+        nb_json = self.get_contents(self.session.path)['content']
+        json_body = {"type": "notebook", "content": nb_json}
+
+        target_url = f"{self.publish_endpoint}{self.session.path}"
```
```suggestion
        target_url = f"{self.publish_endpoint}{path}"
```

willingc commented on bookstore/client/bookstore.py:
```diff
@@ -4,18 +4,35 @@
+    def clone(self, s3_bucket="", s3_key="", target_path=""):
+        s3_bucket = s3_bucket or self.default_bucket
+        json_body = {"s3_bucket": s3_bucket, "s3_key": s3_key, "target_path": target_path}
+        target_url = f"{self.clone_endpoint}"
```
Better style to not create the target_url variable which is only used in the next line. Instead put `f"{self.clone_endpoint}" in the post request.

willingc commented on bookstore/client/bookstore.py:
```diff
@@ -4,18 +4,35 @@
-        target_url = f"{self.publish_endpoint}{self.notebook['path']}"
+    @property
+    def clone_endpoint(self):
+        api_endpoint = "/api/bookstore/cloned/"
```
cloned feels a bit off on the api to me. I would prefer a noun: clone, copy, content...

willingc commented on bookstore/client/notebook.py:
```diff
@@ -182,12 +204,11 @@ def __init__(self):
     @property
     def kernel_id(self):
-        return os.path.basename(self.connection_file).lstrip('kernel-').rstrip('.json')
+        return extract_kernel_id(self.connection_file)
```
Do we anticipate that this function (extract_kernel_id) will be used more than once? If not, it may be better to leave as is.

willingc commented on bookstore/clone.py:
```diff
@@ -0,0 +1,112 @@
+        """
+        s3_bucket = self.get_argument("s3_bucket", "")
+        # s3_paths module has an s3_key function; file_key avoids confusion
+        file_key = self.get_argument("s3_key", "")
```
Perhaps `s3_file_key`

willingc commented on bookstore/publish.py:
```diff
@@ -0,0 +1,80 @@
+        """Publish notebook model to the path"""
+        if model['type'] != 'notebook':
+            raise web.HTTPError(400, "bookstore only publishes notebooks")
+        content = model['content']
```
I would skip this assignment and use `model['content']` when doing the put later.

willingc commented on bookstore/__init__.py:
```diff
@@ -1,3 +1,9 @@
+import os
```
```suggestion
```

willingc commented:
I'm going to go ahead and merge this as is. I'll open a followup PR to address any outstanding feedback. (I'm bored with pneumonia and need something to do.)

willingc commented on CHANGELOG.md:
```diff
@@ -1,10 +1,35 @@
+
+### Significant changes
+
+Validation information is now exposed as a dict at the /api/bookstore endpoint.
```
See PR #87

willingc commented on CHANGELOG.md:
```diff
@@ -1,10 +1,35 @@
+
+[2.2.0](https://github.com/nteract/bookstore/releases/tag/2.2.0)
 
 [Unreleased](https://github.com/nteract/bookstore/compare/0.2.1...HEAD)
```
See PR #87

MSeal commented:
@willingc I think we maybe shouldn't have merged as is. I'd of preferred some tests and author responses to a few of the comments in the PR.

willingc commented:
Sorry, @MSeal. If you or @mpacer feel strongly about this, please feel free to revert.
